src/Access/EnabledQuota.h

```diff
@@ -91,11 +90,12 @@ class EnabledQuota : public boost::noncopyable
 
         std::optional<QuotaUsage> getUsage(std::chrono::system_clock::time_point current_time) const;
     };
+    using IntervalsPtr = std::shared_ptr<const Intervals>;
 
     struct Impl;
 
     const Params params;
-    boost::atomic_shared_ptr<const Intervals> intervals; /// atomically changed by QuotaUsageManager
+    IntervalsPtr intervals; // atomically changed by QuotaCache
```

Comment:
It seems as important change. Be sure there is not date races. `shared_ptr` is not thread safe.

Reply:
It is replaced with `std::atomic_exchange`, so should be safe

Reply:
> It is replaced with std::atomic_exchange, so should be safe
How does it apply here?

https://en.cppreference.com/w/cpp/atomic/atomic_exchange
> Replaces the value of an atomic object with a specified value.

I see you just replaced the atomic object with the ordinary one. So I do not understand how it could be correct. Please explain.

Reply:
Yes, it lacks of `atomic_load`.
Anyway, I've tested various locks, and `boost::atomic_shared_ptr` (with a spilock, that sleeps 1000ns on contention) better then the others - https://gist.github.com/azat/51a5fcc3a40af9f678906a3a6e14e079#file-test-locks-sh